WIP: Adds calenders, Adds frequenices

raftx24 opened the pull request:
adds calendars.
adds

raftx24 committed:
wip

raftx24 committed:
add calendars.
add frequencies

raftx24 committed:
Apply fixes from StyleCI

raftx24 committed:
Merge pull request #1 from raftx24/analysis-qMAOpv

Apply fixes from StyleCI

aocneanu commented on src/AppServiceProvider.php:
```diff
@@ -5,14 +5,16 @@
 
 class AppServiceProvider extends ServiceProvider
 {
     protected $resolvers = [];
```
probably a leftover

aocneanu commented on src/app/Contracts/Calendar.php:
```diff
@@ -0,0 +1,19 @@
+
+interface Calendar
+{
+    public static function getEvents(Request $request): Collection;
```
let's drop the `get` prefix

aocneanu commented on src/app/Contracts/Calendar.php:
```diff
@@ -0,0 +1,19 @@
+{
+    public static function getEvents(Request $request): Collection;
+
+    public function getKey();
```
string

raftx24 replied:
getKey exists in Model too, and if we put string here it conflicts with Model.

aocneanu replied:
right

aocneanu commented on src/app/Contracts/Calendar.php:
```diff
@@ -0,0 +1,19 @@
+
+    public function getKey();
+
+    public function name();
```
string

aocneanu commented on src/app/Contracts/Calendar.php:
```diff
@@ -0,0 +1,19 @@
+
+    public function name();
+
+    public function color();
```
string

aocneanu commented on src/app/Contracts/Calendar.php:
```diff
@@ -0,0 +1,19 @@
+
+    public function color();
+
+    public function readonly();
```
bool?

aocneanu commented on src/app/Contracts/ProvidesEvent.php:
```diff
@@ -18,7 +18,7 @@ public function end(): Carbon;
     public function location();
 
-    public function calendar();
+    public function getCalendar();
```
let's drop `get`. Do we have a `Calendar` return type here?

raftx24 replied:
event has a relation with calendars table, and calendar() conflicts with relation

aocneanu commented on src/app/Contracts/RoutableEvent.php:
```diff
@@ -0,0 +1,10 @@
+
+interface RoutableEvent
+{
+    public function route() :Route;
```
`(): Route`

aocneanu commented on src/app/Http/Requests/ValidateEventUpdate.php:
```diff
@@ -4,6 +4,15 @@
 
 class ValidateEventUpdate extends ValidateEventStore
```
see if you can consolidate the two requests without making compromises. If not, leave it as it is

raftx24 replied:
I tried to do it, but I didn't succeed, because Update has extra rules

raftx24 committed:
refactor

raftx24 committed:
fix bug

raftx24 committed:
fix bug

raftx24 committed:
fix bug

aocneanu committed:
fixes migrations

raftx24 committed:
fix bug

raftx24 committed:
Merge remote-tracking branch 'origin/add_calenders' into add_calenders

raftx24 committed:
fix styleci

aocneanu committed:
refactor

aocneanu committed:
Merge branch 'add_calenders' of https://github.com/raftx24/calendar into add_calenders

aocneanu committed:
makes only work with null

raftx24 committed:
wip

raftx24 committed:
seprate date and time in starts_at and ends_at

raftx24 committed:
Apply fixes from StyleCI

raftx24 committed:
Merge pull request #2 from raftx24/analysis-Xp3xbV

Apply fixes from StyleCI

raftx24 committed:
fix bug

raftx24 committed:
Merge remote-tracking branch 'origin/add_calenders' into add_calenders

gandesc commented on src/AppServiceProvider.php:
```diff
@@ -32,23 +32,13 @@ private function loadDependencies()
         return $this;
     }
 
     private function publisDependencies()
```
method name typo
```suggestion
    private function publishDependencies()
```

gandesc commented on src/app/Forms/Builders/EventForm.php:
```diff
@@ -31,25 +30,36 @@ public function create()
+            ->value('update_type', UpdateType::All)
+            ->value('starts_time', date('H:i', strtotime($event->starts_time)))
+            ->value('ends_time', date('H:i', strtotime($event->ends_time)))
+            ->value('update_type', UpdateType::Single)
```
typo, `->value('update_type'` is also above at line 37

gandesc commented on src/app/Http/Controllers/Calendar/Store.php:
```diff
@@ -0,0 +1,22 @@
+{
+    public function __invoke(ValidateCalendarRequest $request, Calendar $calendar)
+    {
+        tap($calendar)->fill($request->validated())
```
I think we no longer need `tap()` for before filling, as `fill()` returns the model

gandesc commented on src/app/Models/Event.php:
```diff
@@ -6,103 +6,129 @@
     {
-        $this->fillDateAttribute('ends_at', $value, 'Y-m-d H:i:s');
+        $this->fillDateAttribute(
+            'ends_on', $value, config('enso.config.dateFormat')
```
`config('enso.config.dateFormat')` is not needed here, since that's what is used by default if  not given

gandesc commented on src/app/Models/Event.php:
```diff
@@ -6,103 +6,129 @@
     {
-        $this->fillDateAttribute('recurrence_ends_at', $value, 'Y-m-d H:i:s');
+        $this->fillDateAttribute(
+            'recurrence_ends_at', $value, config('enso.config.dateFormat')
```
same as above

gandesc commented on src/app/Models/Event.php:
```diff
@@ -112,69 +138,78 @@ public function recurrenceEnds(): ?Carbon
-                );
-            });
-        });
+        tap($this)->fill($attributes)->save();
```
tap() not required

gandesc commented on src/app/Http/Controllers/Calendar/Index.php:
```diff
@@ -3,12 +3,13 @@
     public function __invoke()
     {
-        return ['calendars' => Calendars::select()];
+        return Calendar::collection(Calendars::all()->values());
```
values() is no longer required

gandesc commented on src/app/Enums/UpdateType.php:
```diff
@@ -0,0 +1,12 @@
+class UpdateType extends Enum
+{
+    const Single = 'single';
+    //const Following = 'following';
```
let's use a better name than following, say 'Futures'

gandesc commented on src/app/Forms/Builders/EventForm.php:
```diff
@@ -31,25 +30,36 @@ public function create()
         return $this->form->value('attendees', $event->attendeeList())
             ->meta('recurrence_ends_at', 'hidden', $event->frequence === Frequencies::Once)
+            ->meta('update_type', 'hidden', $event->frequence === Frequencies::Once)
             ->value('reminders', $this->reminders($event))
```
instead of having this method `reminders()` we should use a Resource for events

gandesc commented on src/app/Http/Controllers/Event/Destroy.php:
```diff
@@ -4,14 +4,17 @@
         $this->authorize('handle', $event);
 
-        $event->delete();
+        tap($event)->deleteEvent($updateType);
```
we no longer need `tap` here

gandesc commented on src/app/Http/Controllers/Event/Store.php:
```diff
@@ -4,24 +4,16 @@
 class Store extends Controller
 {
-    public function __invoke(ValidateEventStore $request, Event $event)
+    public function __invoke(ValidateEventRequest $request, Event $event)
```
when creating an Event we want to make sure the user has access to that calendar.
We can either use a policy in the controller or make use of the `authorize()` method in the request validation class

gandesc commented on src/app/Http/Requests/ValidateEventIndexRequest.php:
```diff
@@ -0,0 +1,23 @@
+use Illuminate\Foundation\Http\FormRequest;
+use LaravelEnso\Calendar\app\Facades\Calendars;
+
+class ValidateEventIndexRequest extends FormRequest
```
either here or in the Controller where we're using the request we should, again, ensure that the user has access to all the calendars he's fetching events for.

gandesc commented on src/app/Services/Frequency/Update.php:
```diff
@@ -0,0 +1,68 @@
+        return $this;
+    }
+
+    protected function deleteOutside()
```
we should not delete past events = events that have already taken place,
so we should exclude those.

gandesc commented on src/app/Services/Frequency/Update.php:
```diff
@@ -0,0 +1,68 @@
+    {
+        if ($this->event->wasChanged(['starts_on', 'recurrence_ends_at'])) {
+            Event::whereParentId($this->parent()->id)->where(function ($query) {
+                $query->where('starts_on', '<=', $this->parent()->starts_on)
```
do we need `smaller than or equals` vs just `smaller` in here?

gandesc commented on src/app/Services/Frequency/Delete.php:
```diff
@@ -0,0 +1,38 @@
+
+    protected function changeParent()
+    {
+        $nextEventId = $this->parent()->events->first()->id;
```
we should order by id or date to ensure the result is as we would expect it.

also, the nextEvent would be the next event after the deleted event, otherwise there could be cases where the retrieved event is something before the deleted one

gandesc commented on src/app/Services/Frequency/Delete.php:
```diff
@@ -0,0 +1,38 @@
+        $nextEventId = $this->parent()->events->first()->id;
+
+        Event::whereParentId($this->parent()->id)->update([
+            'parent_id' => DB::raw("IF(id = {$nextEventId},NULL,{$nextEventId})"),
```
we should only update the future ones, past the deleted event.

gandesc commented on src/app/Traits/HasEvent.php:
```diff
@@ -7,7 +7,7 @@
 
 trait HasEvent
```
I think this trait is not really necessary, Adrian was suggesting using a local Resource if/when we'd want to create custom calendar from Models, rather than using the trait on the models

gandesc commented on composer.json:
```diff
@@ -34,8 +34,8 @@
         "laravel": {
```
we should also add  a `UserServiceProvider` that uses `dynamicMethods` so that we can get a user's events and calendars. The method/relationship for events should be `calendarEvents` so that it's more specific than some event model we could have in an app.

gandesc commented on src/app/Services/Calendars.php:
```diff
@@ -0,0 +1,66 @@
+        return $this->calendars;
+    }
+
+    public function only(?array $calendars)
```
no `?`, the parameter should not be nullable

gandesc commented on src/database/migrations/2019_10_16_100000_create_calendars_table.php:
```diff
@@ -0,0 +1,40 @@
+            $table->timestamps();
+        });
+
+        Schema::table('events', function (Blueprint $table) {
```
the calendars migration should run before the calendar_events, so that we wouldn't need to make these updates due to the migrations' order

gandesc commented on src/database/migrations/2019_03_21_100000_create_calendar_events_table.php:
```diff
@@ -4,20 +4,27 @@
 
-            $table->datetime('starts_at');
-            $table->datetime('ends_at');
+            $table->date('starts_on')->index();
```
let's go with `starts/ends_time` instead of `starts/ends_on`

gandesc commented on src/app/Http/Requests/ValidateEventRequest.php:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    protected function genericRule()
```
let's rename this to `requiredOrFilled`

raftx24 committed:
fix bugs.
adds future updating
sets the frequency to once temporary

gandesc commented:
@raftx24 in the calendars migration file there's a leftover typo for the migration rollback `Schema::dropIfExists('events');`

raftx24 commented on src/database/migrations/2019_03_21_100000_create_calendar_events_table.php:
```diff
@@ -36,6 +44,6 @@ public function up()
     public function down()
     {
-        Schema::dropIfExists('events');
+        Schema::dropIfExists('calendar_events');
```
@gandesc I solved it here.

gandesc committed:
- fixes custom events overwriting native events
- removes unneeded calendar property in the event resource
- adds an optional people Birthday custom calendar implementation

gandesc committed:
removed resource load

gandesc committed:
updated as per styleci notes

gandesc committed:
fixed blank spaces
